server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregationBuilder.java

```diff
@@ -48,42 +49,71 @@
         ValuesSourceParserHelper.declareNumericFields(PARSER, true, true, true);
-
         PARSER.declareInt(AutoDateHistogramAggregationBuilder::setNumBuckets, NUM_BUCKETS_FIELD);
+        PARSER.declareStringOrNull(AutoDateHistogramAggregationBuilder::setMinimumIntervalExpression, MINIMUM_INTERVAL_FIELD);
     }
 
+    public static final Map<Rounding.DateTimeUnit, String> allowedIntervals = Map.ofEntries(
+        entry(Rounding.DateTimeUnit.YEAR_OF_CENTURY, "year"),
+        entry(Rounding.DateTimeUnit.MONTH_OF_YEAR, "month"),
+        entry(Rounding.DateTimeUnit.DAY_OF_MONTH, "day"),
+        entry( Rounding.DateTimeUnit.HOUR_OF_DAY, "hour"),
+        entry(Rounding.DateTimeUnit.MINUTES_OF_HOUR, "minute"),
+        entry(Rounding.DateTimeUnit.SECOND_OF_MINUTE, "second")
+    );
+
     /**
      *
      * Build roundings, computed dynamically as roundings are time zone dependent.
      * The current implementation probably should not be invoked in a tight loop.
      * @return Array of RoundingInfo
      */
-    static RoundingInfo[] buildRoundings(ZoneId timeZone) {
+    static RoundingInfo[] buildRoundings(ZoneId timeZone, String minimumInterval) {
+
+        int indexToSliceFrom = 0;
+
         RoundingInfo[] roundings = new RoundingInfo[6];
         roundings[0] = new RoundingInfo(createRounding(Rounding.DateTimeUnit.SECOND_OF_MINUTE, timeZone),
-            1000L, "s", 1, 5, 10, 30);
+            1000L, "s",
+            allowedIntervals.get(Rounding.DateTimeUnit.SECOND_OF_MINUTE),1, 5, 10, 30);
         roundings[1] = new RoundingInfo(createRounding(Rounding.DateTimeUnit.MINUTES_OF_HOUR, timeZone),
-            60 * 1000L, "m", 1, 5, 10, 30);
+            60 * 1000L, "m",
+            allowedIntervals.get(Rounding.DateTimeUnit.MINUTES_OF_HOUR), 1, 5, 10, 30);
         roundings[2] = new RoundingInfo(createRounding(Rounding.DateTimeUnit.HOUR_OF_DAY, timeZone),
-            60 * 60 * 1000L, "h",1, 3, 12);
+            60 * 60 * 1000L, "h",
+            allowedIntervals.get(Rounding.DateTimeUnit.HOUR_OF_DAY), 1, 3, 12);
         roundings[3] = new RoundingInfo(createRounding(Rounding.DateTimeUnit.DAY_OF_MONTH, timeZone),
-            24 * 60 * 60 * 1000L, "d", 1, 7);
+            24 * 60 * 60 * 1000L, "d",
+            allowedIntervals.get(Rounding.DateTimeUnit.DAY_OF_MONTH), 1, 7);
         roundings[4] = new RoundingInfo(createRounding(Rounding.DateTimeUnit.MONTH_OF_YEAR, timeZone),
-            30 * 24 * 60 * 60 * 1000L, "M", 1, 3);
+            30 * 24 * 60 * 60 * 1000L, "M",
+            allowedIntervals.get(Rounding.DateTimeUnit.MONTH_OF_YEAR), 1, 3);
         roundings[5] = new RoundingInfo(createRounding(Rounding.DateTimeUnit.YEAR_OF_CENTURY, timeZone),
-            365 * 24 * 60 * 60 * 1000L, "y", 1, 5, 10, 20, 50, 100);
-        return roundings;
+            365 * 24 * 60 * 60 * 1000L, "y",
+            allowedIntervals.get(Rounding.DateTimeUnit.YEAR_OF_CENTURY), 1, 5, 10, 20, 50, 100);
```

Comment:
suggestion: instead of having to pass in the same `DateTimeUnit` twice and relying on them being the same we could change `RoundingInfo` so it's constructor has the signature:
```
public RoundingInfo(DateTimeUnit dateTimeUnit,
                            ZoneId timeZone,
                            long roughEstimateDurationMillis,
                            String unitAbbreviation,
                            int... innerIntervals)
```
The `RoundingInfo` constructor could then call `createRounding()` and `allowedIntervals.get()` and this would ensure we can never have a `RoundingInfo` where the `dateTimeUnit` String and the Rounding don't match the same `DateTimeUnit`?

Reply:
great suggestion! this simplified the calling logic a lot. addressed in 95ee1cf0320a901e70efa41a3c9f9785d5aeb59f